hivemind/p2p/p2p_daemon.py

```diff
@@ -0,0 +1,326 @@
+                writer.close()
+                return
+            try:
+                result = handle(request)
+                await P2P.send_data(result, writer)
+            except Exception as exc:
+                await P2P.send_data(exc, writer)
+            finally:
+                writer.close()
+
+        return do_handle_stream
+
+    @staticmethod
+    def _handle_unary_stream(handle, context, in_proto_type, out_proto_type):
+        async def watchdog(reader: asyncio.StreamReader):
+            await reader.read(n=1)
+            raise P2P.InterruptedError()
+
+        async def do_handle_unary_stream(
+                stream_info: StreamInfo,
+                reader: asyncio.StreamReader,
+                writer: asyncio.StreamWriter) -> None:
+            try:
+                try:
+                    request = await P2P.receive_protobuf(in_proto_type, reader)
+                except asyncio.IncompleteReadError:
+                    logger.debug("Incomplete read while receiving request from peer")
+                    return
+                except google.protobuf.message.DecodeError as error:
+                    logger.warning(repr(error))
+                    return
+
+                context.peer_id, context.peer_addr = stream_info.peer_id, stream_info.addr
+                done, pending = await asyncio.wait([watchdog(reader), handle(request, context)],
+                                                   return_when=asyncio.FIRST_COMPLETED)
+                try:
+                    result = done.pop().result()
+                    await P2P.send_protobuf(result, out_proto_type, writer)
+                except P2P.InterruptedError:
+                    pass
+                except Exception as exc:
+                    await P2P.send_data(exc, writer)
+                finally:
+                    pending_task = pending.pop()
+                    pending_task.cancel()
+                    try:
+                        await pending_task
+                    except asyncio.CancelledError:
+                        pass
+            finally:
+                writer.close()
+
+        return do_handle_unary_stream
+
+    def start_listening(self):
+        async def listen():
+            async with self._client.listen():
+                await self._server_stopped.wait()
+
+        self._listen_task = asyncio.create_task(listen())
+
+    async def stop_listening(self):
+        if self._listen_task is not None:
+            self._server_stopped.set()
+            self._listen_task.cancel()
+            try:
+                await self._listen_task
+            except asyncio.CancelledError:
+                self._listen_task = None
+                self._server_stopped.clear()
+
+    async def add_stream_handler(self, name, handle):
+        if self._listen_task is None:
+            self.start_listening()
+        await self._client.stream_handler(name, P2P._handle_stream(handle))
```

Comment:
Unlike the unary handlers, the stream handlers use `pickle` for serializing inputs/outputs and lead to the risk of remote code execution (see #155).

While an application is not vulnerable until a stream handler is added (and the DHT-over-libp2p implementation in #165 does not use it), I still think that the presence of such a convenient API may seduce a programmer to use it (and accidentally allow their experiment to be turned into a botnet).

Therefore, I suggest to either remove `P2P.add_stream_handler`, `P2P. _handle_stream`, and other methods using `pickle` from this PR or rewrite them with `MSGPackSerializer`.